### PR TITLE
Add Linux Unified Key Setup (LUKS)

### PR DESCRIPTION
**Pull request by davidhicks**

### CODE REVIEW

File: filesystem/luks.ksy [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        type: key_slot
+        repeat: expr
+        repeat-expr: 8
+    types:
+      key_slot:
+        seq:
+          - id: state_of_key_slot
+            type: u4
+            doc: |
+              Value is 0x0000DEAD if the key slot is disabled.
```
Comment (KOLANICH) [inline filesystem/luks.ksy]: make a enum

Comment (KOLANICH) [inline filesystem/luks.ksy]: is this field limited by this 2 values? If it is, make a enum.

**Comment by davidhicks** [comment]
Thanks @KOLANICH for the review. The key slot state can only be ENABLED or DISABLED, so I've made it an enumeration as suggested.

File: filesystem/luks.ksy [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+          key_material:
+            pos: start_sector_of_key_material * 512
+            size: _parent.number_of_key_bytes * number_of_anti_forensic_stripes
+        enums:
+          key_slot_states:
+            0x0000DEAD: disabled_key_slot
+            0x00AC71F3: enabled_key_slot
+instances:
+  payload:
+    pos: partition_header.payload_offset * 512
```
Comment (KOLANICH) [inline filesystem/luks.ksy]: I'd like to move 512 into an instance ```sector_size```.
Also I wonder if it is a physical sector size (which can be 4k).

Comment (KOLANICH) [reply filesystem/luks.ksy]: OK, thank you for the contribution.

File: filesystem/luks.ksy [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  encryption parameters and up to 8 user keys (which can unlock the master key).
+doc-ref: https://gitlab.com/cryptsetup/cryptsetup/wikis/LUKS-standard/on-disk-format.pdf
+seq:
+  - id: partition_header
+    type: partition_header
+types:
+  partition_header:
+    seq:
+      - id: magic
+        contents: [0x4C, 0x55, 0x4B, 0x53, 0xBA, 0xBE]
```
Comment (KOLANICH) [inline filesystem/luks.ksy]: ```["LUKS", 0xBA, 0xBE]```

File: filesystem/luks.ksy [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      - id: version
+        contents: [0x00, 0x01]
+      - id: cipher_name_specification
```
Comment (KOLANICH) [inline filesystem/luks.ksy]: Look like a new type.
```yaml
spec_str:
  seq:
    - id: str
      type: str
      size: 32
```

Comment (GreyCat) [reply filesystem/luks.ksy]: I wouldn't go as far as creating new type for a mere repetition of same 32-byte string for 3 times.

**Comment by GreyCat** [comment]
@davidhicks, thanks for your contribution! Sorry for getting it so long to review. I'll merge it now, but please take a look at a new notes I'll add :)

File: filesystem/luks.ksy [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            pos: start_sector_of_key_material * 512
+            size: _parent.number_of_key_bytes * number_of_anti_forensic_stripes
+        enums:
+          key_slot_states:
+            0x0000DEAD: disabled_key_slot
+            0x00AC71F3: enabled_key_slot
+instances:
+  payload:
+    pos: partition_header.payload_offset * 512
+    size-eos: true
```
Comment (GreyCat) [inline filesystem/luks.ksy]: Please note that EOF newline is obligatory: http://doc.kaitai.io/ksy_style_guide.html#_general_formatting, I'll add it.

File: filesystem/luks.ksy [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        type: str
+        size: 32
+      - id: cipher_mode_specification
+        type: str
+        size: 32
+      - id: hash_specification
+        type: str
+        size: 32
+      - id: payload_offset
+        type: u4
```
Comment (GreyCat) [inline filesystem/luks.ksy]: If we'll stick to [style guide](http://doc.kaitai.io/ksy_style_guide.html#attr-id), it's recommended to use `ofs_payload` for attributes like this.

File: filesystem/luks.ksy [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      - id: number_of_key_bytes
+        type: u4
```
Comment (GreyCat) [inline filesystem/luks.ksy]: Probably `len_key`, or something like that, if we'll stick to style guide.

File: filesystem/luks.ksy [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            type: u4
+            enum: key_slot_states
+          - id: iteration_parameter
+            type: u4
+          - id: salt_parameter
+            size: 32
+          - id: start_sector_of_key_material
+            type: u4
+          - id: number_of_anti_forensic_stripes
+            type: u4
```
Comment (GreyCat) [inline filesystem/luks.ksy]: Probably something like `num_anti_forensic_stripes`.

**Pull request merged**